applications/main/clock/clock_app.c

```diff
@@ -0,0 +1,197 @@
+    canvas_set_font(canvas, FontSecondary);
+    canvas_draw_str_aligned(canvas, 64, 42, AlignCenter, AlignTop, date_string);
+
+    if(state->settings.time_format == H12)
+        canvas_draw_str_aligned(canvas, 65, 12, AlignCenter, AlignCenter, meridian_string);
+
+    FURI_LOG_T(TAG, "clock_render_callback - end");
+}
+
+static void clock_state_init(ClockState* const state) {
+    bool load_success = LOAD_CLOCK_SETTINGS(&state->settings);
+    if(!load_success) {
+        state->settings = DEFAULT_SETTINGS;
+    }
+    FURI_LOG_I(TAG, "Time format: %s", state->settings.time_format == H12 ? "12h" : "24h");
+    FURI_LOG_I(
+        TAG, "Date format: %s", state->settings.date_format == Iso ? "ISO 8601" : "RFC 5322");
+    furi_hal_rtc_get_datetime(&state->datetime);
+}
+
+// Runs every 1000ms by default
+static void clock_tick(void* ctx) {
+    furi_assert(ctx);
+    FuriMessageQueue* event_queue = ctx;
+    PluginEvent event = {.type = EventTypeTick};
+    // It's OK to loose this event if system overloaded
+    furi_message_queue_put(event_queue, &event, 0);
+}
+
+int32_t clock_app(void* p) {
+    UNUSED(p);
+    ClockState* plugin_state = malloc(sizeof(ClockState));
+
+    plugin_state->event_queue = furi_message_queue_alloc(8, sizeof(PluginEvent));
+    if(plugin_state->event_queue == NULL) {
+        FURI_LOG_E(TAG, "Cannot create event queue");
+        free(plugin_state);
+        return 255;
+    }
+    FURI_LOG_D(TAG, "Event queue created");
+
+    plugin_state->mutex = furi_mutex_alloc(FuriMutexTypeNormal);
+    if(plugin_state->mutex == NULL) {
+        FURI_LOG_E(TAG, "Cannot create mutex");
+        furi_message_queue_free(plugin_state->event_queue);
+        free(plugin_state);
+        return 255;
+    }
+    FURI_LOG_D(TAG, "Mutex created");
+
+    clock_state_init(plugin_state);
+
+    // Set system callbacks
+    ViewPort* view_port = view_port_alloc();
+    view_port_draw_callback_set(view_port, clock_render_callback, plugin_state);
+    view_port_input_callback_set(view_port, clock_input_callback, plugin_state->event_queue);
+
+    FuriTimer* timer =
+        furi_timer_alloc(clock_tick, FuriTimerTypePeriodic, plugin_state->event_queue);
+
+    if(timer == NULL) {
+        FURI_LOG_E(TAG, "Cannot create timer");
+        furi_mutex_free(plugin_state->mutex);
+        furi_message_queue_free(plugin_state->event_queue);
+        free(plugin_state);
+        return 255;
+    }
+    FURI_LOG_D(TAG, "Timer created");
+
+    // Open GUI and register view_port
+    Gui* gui = furi_record_open(RECORD_GUI);
+    gui_add_view_port(gui, view_port, GuiLayerFullscreen);
+
+    furi_timer_start(timer, furi_kernel_get_tick_frequency());
+    FURI_LOG_D(TAG, "Timer started");
+
+    // Main loop
+    PluginEvent event;
+    for(bool processing = true; processing;) {
+        FuriStatus event_status = furi_message_queue_get(plugin_state->event_queue, &event, 100);
```

Comment:
Using 100ms timeout make sense only if you do some work on expire, otherwise just put forever.